moabb/datasets/base.py

```diff
@@ -654,3 +656,206 @@ def data_path(
+                ".bdf",
+                ".EDF",
+                ".snirf",
+                ".cdt",
+                ".mef",
+                ".nwb",
+            ],
+        }
+        if subject is not None:
+            out["subjects"] = str(subject)
+        return out
+
+    def _get_read_extra_params(
+        self,
+        subject: int,  # pylint: disable=unused-argument
+    ) -> dict[str, Any] | None:
+        """Return the ``extra_params`` argument for the :func:`mne_bids.read_raw_bids` function."""
+        return None
+
+    @staticmethod
+    def _find_matching_paths(root, **kwargs) -> list[mne_bids.BIDSPath]:
+        bids_paths = mne_bids.find_matching_paths(root=root, **kwargs)
+        # Remove JSON files manually (the ignore_json argument only arrives in mne-bids=0.16)
+        return [bids_path for bids_path in bids_paths if bids_path.extension != ".json"]
+
+    @abc.abstractmethod
+    def _download_subject(self, subject, path, force_update, update_path, verbose) -> str:
+        """Download the data of a single subject and return the local path to the ROOT of the BIDS dataset.
+
+        Returns
+        -------
+        root : str
+            Path to the ROOT of the BIDS dataset.
+        """
+        pass
+
+    def bids_paths(
+        self, subject, path=None, force_update=False, update_path=None, verbose=None
+    ) -> list[mne_bids.BIDSPath]:
+        root = self._download_subject(subject, path, force_update, update_path, verbose)
+        return self._find_matching_paths(
+            root=root, **self._get_path_search_params(subject)
+        )
+
+    def data_path(
+        self, subject, path=None, force_update=False, update_path=None, verbose=None
+    ):
+        bids_paths = self.bids_paths(subject, path, force_update, update_path, verbose)
+        return [bids_path.fpath for bids_path in bids_paths]
+
+    def _get_single_subject_data(self, subject):
+        bids_paths = self.bids_paths(subject)
+        data = {}
+        for bids_path in bids_paths:
+            raw = mne_bids.read_raw_bids(
+                bids_path, extra_params=self._get_read_extra_params(subject)
+            )
+            # Data needs to be preloaded for the filtering step of paradigms
+            raw.load_data()
+
+            if bids_path.session is None:
+                log.warning(
+                    "Session not found for subject='%s'. Using session='0'", subject
+                )
+                session = "0"
+            else:
+                session = bids_path.session
+            if bids_path.run is None:
+                log.warning(
+                    "Run not found for subject='%s', session='%s'. Using run='0'",
+                    subject,
+                    session,
+                )
+                run = "0"
+            else:
+                run = bids_path.run
+            data.setdefault(session, {})[run] = raw
+        return data
+
+
+class LocalBIDSDataset(BaseBIDSDataset):
```

Comment:
I would love to make one osfdataset with this...

Reply:
Maybe instead of `LocalBIDSDataset`,  you can create an `OsfDataset` class which inherits from `BaseBIDSDataset` and directly implements the `_download_subject` method using `osfclient` or another API.